wip (#93)
NixOS/rfcs: README.md

# Nix RFCs (Request For Comments)

Many ideas, including bug fixes and documentation improvements can be
implemented and reviewed via the normal GitHub pull request workflow.

> Some ideas though are "substantial", and we ask that these be put through a

Reviewer: Not sure why we're replacing changes with ideas throughout this document. This seems like a potentially contentious choice, with little upside, so I'd rather leave this out.

Author: As stated above, probably the main purpose of _informational_ RFCs is to record consensus on something that does not primarily originate in a change.

My best word for that something has been: "idea". Maybe there is a better word?

discussed. The Shepherd Team will propose to either accept or reject the RFC
after the FCP.

##### RFC Categories
In order to do do justice to the different aspects of documents that merit

> generation of broad community consensus via the RFC process, we cassify each

Reviewer 2: ```suggestion
generation of broad community consensus via the RFC process, we classify each
```

RFC into _feature_, _process_ and _informational_ RFCs. All follow the same
high-level process as described above, but each category requires a different

> "mode of discussion", templates, critarias and judgment that it is benefical

Reviewer 2: ```suggestion
"mode of discussion", templates, criteria and judgment that it is beneficial
```

to the overall RFC process to identify those categories explicitly.

## Process from Creation to Merge

> *In short, to get a major change included in Nix, Nixpkgs or the Ecosystem, one must

Reviewer 2: ```suggestion
*In short, to get a major change included in Nix, Nixpkgs or the ecosystem, one must
```
maybe? I'm not a native speaker. Also I don't know if I even should propose typo fixes or if you wanted to read over it anyways.

first get the RFC merged into the RFC repository as a markdown file under the
corresponding directory. At that point the RFC is accepted and may be implemented

> with the goal of eventual inclusion into Nix, Nixpkgs or the Ecosystem.*

Reviewer 2: ```suggestion
with the goal of eventual inclusion into Nix, Nixpkgs or the Ecosystem.
```
or is it referenced later or so?

> 0. Have a cool idea r an important information!

Reviewer 2: ```suggestion
0. Have a cool idea or an important information!
```

> 1. Identify its category: _feature_, _process_ or _informational_.

Reviewer 3: Where is the guidance for picking the category? It appears to be at the start of the template so that should be specified so that a potential author knows where to look.

> 2. Start with the correct template and follow the instructions and comments.

Reviewer 3: Please specify where the templates live.

3. Submit a pull request. As a pull request the RFC will receive design feedback
   from the larger community, and the author should be prepared to revise it in

>    response. Leave them in draft while you're still actively working on them.

Reviewer: This one is a bit contentious too IMO. If RFC stands for Request For Comments, then it should be implied that you're requesting feedback, and that it's not necessarily fully done the moment you post it. So why create draft RFCs at all?

Author: There is currently a draft label to denote the draft stage. This is indeed orthogonal to this RFC, but I thought it might help the RFC steering commitee once things scale up (if they do).

I'll remove this as a strictly unrelated change.

![RFC Process](./rfcs/0036-rfc-process.png)
![Review Process](./rfcs/0036-review-process.png)

### Tips

> ###### Create Prototypes

Reviewer: ```suggestion
#### Create Prototypes
```

In case your RFC is a technical proposal, you might want to prepare a
prototype of your idea to firstly make yourself aware of potential pitfalls
and also help reviewers understand the RFC. Code may be able to explain some
issues in short.

###### Motivation First
In any case, it can be a good idea to elaborate the motivation first, while
leaving the PR in draft. This helps to ensure, that when the solution is
discussed in the future, interested participants can have a clear and refined

> picture of the problem that you try to address.

Reviewer: I disagree on this one, as mentioned above.

Also, why is this RFC touching so many parts of the RFC process? It would be better if it only focused on adding the concept of categories, IMO.

Author: Some time ago, I acknowledged that two stage discussions are usually more informed and productive.

In my experience, it tremendously helps with forming consensus, when first focusing on the motives and then, after a while, proceeding to the implementation.

If the stages are not separated enough, discussion can (stressfully for all participants) take place on different levels and on different spots on the causality vector. In short: convoluted.

Reviewer 3: I am not opposed to this addition but in the interest of keeping the RFC as small as possible it may be best to defer this or propose it as a separate RFC.

## The RFC life-cycle

Once an RFC is accepted the authors may implement it and for example submit the
feature as a pull request to the Nix or Nixpkgs repo. Being accepted is not a
rubber stamp, and in particular still does not mean the feature will ultimately
be merged; it does mean that in principle all the major stakeholders have agreed
to the feature and are amenable to merging it. In general though this means that
the implementation will be merged as long as there are no substantial technical

> objections to the implementation.

Reviewer 3: Maybe `s/feature/change/`? It may also make sense to indicate that not all RFCs prompt change. How about:

> Once an RFC is accepted it may be implemented if any further work is required. This may be via a pull request to the nix or Nixpkgs repo or via other means. Being accepted is not a rubber stamp, and in particular still does not mean the change will ultimately be enacted; it does mean that in principle all the major stakeholders have agreed to the
change and are amenable to its implementation as long as there are no substantial technical objections to the implementation.

Furthermore, the fact that a given RFC has been accepted implies nothing about
what priority is assigned to its implementation, nor does it imply anything
about whether a Nix/Nixpkgs community member has been assigned the task of
implementing the RFC. While it is not necessary that the author of the RFC also
does the implementation, it is by far the most effective way to see an RFC through to
completion: authors should not expect that other project developers will take on
responsibility for implementing their accepted feature.

Minor modifications to accepted RFCs can be done in follow-up pull requests. We

> strive to write each RFC in a manner that it will  reflect the final state of

Reviewer 3: ```suggestion
strive to write each RFC in a manner that it will reflect the final state of
```
